### ISSUE TITLE
[Tracking] Add a reportActionCount field to reports in Onyx

### ISSUE THREAD

**Issue by JmillsExpensify** [Expensify/App#15352]
What follows below was originally posted by @roryabraham [here](https://github.com/Expensify/App/issues/14523#issuecomment-1439269331). This issue is intended to track what's need to resolve the odd behavior that https://github.com/Expensify/App/issues/14523 was originally meant to solve, which resulted in a regression and requires the more holistic solution below. 

-----
# Proposal
1. Start maintaining an accurate `reportActionCount` field
1. Don't show DM's with only 1 reportAction (i.e: the `CREATED` action) in options list, unless it is the currently viewed report

# Please re-state the problem that we are trying to solve in this issue.
When you create a new DM but don't send any messages, it should not appear in the LHN for other users. If you don't create a draft message, then navigate away from the chat, it should disappear from the LHN for you.

# What is the root cause of that problem?
(after https://github.com/Expensify/App/pull/15334 is merged) There is no special logic in `ReportUtils.shouldReportBeInOptionsList` to hide DMs with only a created action from the LHN – the default is to show the report.

# What changes do you think we should make in order to solve the problem?

## Add a `reportActionCount` field to reports in Onyx

To make this change, we need to examine all NewDot flows that add or remove reportActions from a report. This includes:

- Creating a new report (adds a `CREATED` action)
- Sending a message or attachment
- Sending an IOU
- Deleting a reportAction

### Propagating `reportActionCount` correctly in the API

First, we need to make sure that when we initially fetch reports, those reports contain their canonical `reportActionCount` from the back-end. Fortunately, the `reportActionCount` is already returned from `Auth::GetChats`, `Auth::GetReportSummaryList`, `Auth::CreateChatReports`, etc..., so all we have to do for this is to update [`ReportUtils::structureReportForOnyx`](https://github.com/Expensify/Web-Expensify/blob/12106f5130cb4f0883675d1129a398f12b7dcdab/lib/ReportUtils.php#L2418) to include `reportActionCount`, which, as @PauloGasparSv pointed out above, is already used to generate the `maxSequenceNumber` field.

Next, we'll need to update `ReportUtils::buildOnyxDataForNewAction` [here](https://github.com/Expensify/Web-Expensify/blob/12106f5130cb4f0883675d1129a398f12b7dcdab/lib/ReportUtils.php#L1815-L1829) so that we include `reportActionCount` as well as the other fields.

Lastly, we'll update [`ReportAPI::editComment`](https://github.com/Expensify/Web-Expensify/blob/41873abd533de094c0e38387e6689b17c03e4548/lib/ReportAPI.php#L492-L623), so that, if we're deleting a comment we always update the report in Onyx to decrement the `reportActionCount`, regardless for whether it's the newest visible reportAction on the report, or if we `shouldKeyReportActionsByID`.

#### Creating a new report
There are a few locations where new reports are created in the NewDot front-end, and we can find them by searching for usages of `ReportUtils.buildOptimisticChatReport`:

- In [`src/libs/actions/Report.navigateToAndOpenReport`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L411)
- In [`src/libs/actions/Report.addPolicyReport`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L959)
- In [`ReportUtils.buildOptimisticWorkspaceChats`](https://github.com/Expensify/App/blob/a5b9c9ba88fa0634a963bd248152b79e3006bd6d/src/libs/ReportUtils.js#L1134)
- In [`IOU.requestMoney`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/IOU.js#L63)
- In [`IOU.createSplitsAndOnyxData`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/IOU.js#L232)
- In [`IOU.getSendMoneyParams`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/IOU.js#L676)

In each of the above-listed locations, there is some code to add an optimistic `CREATED` action, along with potentially more actions such as an IOU action. We will want to trace each of these flows and ensure that we set the correct optimistic `reportActionCount` according to the number of optimistically-generated `reportActions` before writing the report to Onyx.

- The optimistic `reportActionCount` can safely be assumed to be `1` in each of these places:
    - [Here in `openReport`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L373-L395)
    - [here in `addPolicyReport`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L956-L1027)
    - For each optimistic report in [`buildOptimisticWorkspaceChats`](https://github.com/Expensify/App/blob/a5b9c9ba88fa0634a963bd248152b79e3006bd6d/src/libs/ReportUtils.js#L1134)
- In `IOU.requestMoney` and `IOU.getSendMoneyParams`, the optimistic `reportActionCount` should be _incremented_ by `1` if we are not creating a new report, but should be `2` if we are creating a new report
- `IOU.createSplitsAndOnyxData` should use the same defaults as `IOU.requestMoney`, except it will vary for the group DM and for every 1:1 DM associated with the bill split.

#### Sending a message / attachment
This happens in `Report.addActions`, where we'll always be able to assume that the report has an existing `reportActionCount` field. We will increment that field by `1` if there is just text or just an attachment, but `2` if there is both text and an attachment.

#### Sending an IOU
This is covered by the same flows listed above under `Creating a new report`. If creating a new report, the correct value for `reportActionCount` on the new report is `2`. If adding an `IOU` action to an existing report, the `reportActionCount` should be incremented by `1`.

#### Deleting a reportAction
This flow lives in [`src/libs/actions/Report.deleteReportComment`](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L678-L751). The report is optimistically updated [here](https://github.com/Expensify/App/blob/5f0921d2f011ccc159870cbad126da0dff9e9f5c/src/libs/actions/Report.js#L696-L701), so we need to update that piece of code to also optimistically decrement the `reportActionCount` on the report by `1`.

## Don't show DM's with only 1 reportAction in options list (unless it is the currently viewed report)

To do this, we will draw inspiration from https://github.com/Expensify/App/pull/15129: 

1. Add a function called `isDirectMessage` to `src/libs/ReportUtils` that returns `true` if the `chatType` rNVP on the report is empty, and `false` if it is not.
1. Update `src/libs/ReportUtils.shouldReportBeInOptionList` to return `false` if the report is a direct message and the `reportActionCount` is `1`.

# Rollout plan
 - https://github.com/Expensify/App/issues/15353
 - https://github.com/Expensify/App/issues/15354
 - https://github.com/Expensify/App/issues/15355
 - https://github.com/Expensify/App/issues/15356
 - https://github.com/Expensify/App/issues/15357
 - https://github.com/Expensify/App/issues/15358
 - https://github.com/Expensify/App/issues/15359
 - https://github.com/Expensify/App/issues/15360
 - https://github.com/Expensify/App/issues/15362
 - https://github.com/Expensify/App/issues/15363

### Issues on hold

- https://github.com/Expensify/App/issues/15966

**Comment by cead22** [comment]
I can help with query changes and questions, and feel free to throw any other questions my way, but I don't have cycles to take the lead on this, at least not until after ECX

**Comment by PauloGasparSv** [comment]
Nice!! I'll create an extra issue to do the Auth update on `GetChats` and assign to myself I think (and ping if I get blocked) ASAP.

Thks @roryabraham @cead22

**Comment by PauloGasparSv** [comment]
Updates coming this week.

**Comment by hayata-suenaga** [comment]
I'd like to take some of the tasks in this rollout plan once the PR on the server side is done

**Comment by PauloGasparSv** [comment]
I'll try to finish the first issue from this by tomorrow.

**Comment by JmillsExpensify** [comment]
Looks like you're still working through that first issue? Do you expect to finish up this week?

**Comment by PauloGasparSv** [comment]
> Looks like you're still working through that first issue? Do you expect to finish up this week?

Planning on finishing it today/tomorrow @JmillsExpensify!

**Comment by PauloGasparSv** [comment]
[Latest discussion](https://expensify.slack.com/archives/C03TQ48KC/p1680526197723869) we had on this last week may lead to scope of changes here:

- We discussed other solutions that don't involve adding the `reportActionCount` like for example using the `lastMessageHtml` field to check if the chat had any previous messages.

- We also discussed creating an index for the `reportActionCount` query, but not everyone agreed on that.

I don't think we reached an agreement so I'll send a new message in the thread to confirm with the team how everyone thinks we should approach this **later Today**.

**Comment by PauloGasparSv** [comment]
Following [last week's discussion](https://expensify.slack.com/archives/C03TQ48KC/p1681158468298129?thread_ts=1680526197.723869&cid=C03TQ48KC) I'll try to make the feature work in https://github.com/Expensify/App/issues/14523 using only the `lastMessageHtml` field.

Keeping this tracking issue open until a working P.R. for that is merged [in case we need to fallback](https://expensify.slack.com/archives/C03TQ48KC/p1681209073265499?thread_ts=1680526197.723869&cid=C03TQ48KC) to the `reportActionCount`

**Comment by PauloGasparSv** [comment]
Latest update [here](https://github.com/Expensify/App/issues/14523#issuecomment-1524054762)

**Comment by PauloGasparSv** [comment]
Latest update [here](https://github.com/Expensify/App/issues/14523#issuecomment-1539084163)

**Comment by PauloGasparSv** [comment]
Latest update [here](https://github.com/Expensify/App/issues/14523#issuecomment-1552012535)

**Comment by PauloGasparSv** [comment]
No updates since [this](https://github.com/Expensify/App/issues/14523#issuecomment-1561879588)

**Comment by PauloGasparSv** [comment]
Still in progress, latest update [here](https://github.com/Expensify/App/issues/14523#issuecomment-1576692880)

**Comment by PauloGasparSv** [comment]
I shifted focus to Manual Requests for a bit but I'm back and making progress Today.

Latest updates explaining the state I'm right now (since the last merge from `main` the LNH logic changed and the P.R. stopped working, fixing that) [in P.R.](https://github.com/Expensify/App/pull/19321)

**Comment by hayata-suenaga** [comment]
I'm unassigning myself as the migration issue is taking much of my time

**Comment by roryabraham** [comment]
Sorry, I'm not sure I should be assigned here anymore either since it's been a long time since I've contributed to or commented on this issue.

**Comment by luacmartins** [comment]
I DMed Jason, Rory and Paulo today and I think we can close this tracking issue since we don't seem to need this solution anymore - see [this comment](https://github.com/Expensify/App/issues/15353#issuecomment-1604967708). We can always reopen if needed.
